### DISCUSSION TITLE
Request for "You are already logged in" to redirect to original target URL

### DISCUSSION THREAD

**Discussion by stelewis-redhat** [keycloak/keycloak#11158] [Q&A]
A couple of example complaints, to show what users are experiencing:

> sso.redhat.com loses original URL when user is already logged in
> 
> The workflow where I quite regularly encounter this is:
> 1. Open e-mail in the morning
> 2. While reading new messages, open various links that need some attention later
> 3. Switch to browser to find tabs are redirected to log in form (because I have not logged in yet this day)
> 4. Log in in one of the tags, get redirected to the page I wanted to open.
> 5. Try to log in on another tab → "You are already logged in"
> 6. Now I have to go back to e-mail to find the relevant links again (only this time there are no unread messages that would make it easy).
> 
> Ideal solution would simply skip the error page and take me immediately to what I want to see.
> If that is not possible, can you at least add a link to the "You are already logged in" page that would take me to the URL that redirected me to SSO in the first place?

and:

> If already logged in on sso.redhat.com, the redirection to the original page doesn't happen after authentication. Only a page is displayed which shows "already logged in" with no way to navigate to the original page via the browser back button long click to select the appropriate page in the tab history which is not sso.redhat.com. This is extremely annoying at scale of anything more than 1.

What is the best way to approach this? If the user has N tabs open and they refresh, Keycloak is doing it correctly for the first tab, but the "You are already logged in" pages seem to lose the state. What's the best way to go about handling this? Will this require changes in Keycloak itself? I don't have any solutions myself, and am at the beginning of my investigations, but wanted to start the discussion. Thanks in advance.

**Comment by alechenninger** [comment]
@mposolda any ideas here? A few things come to mind:

1. Have an option per client to redirect to the callback URL even if the user is already authenticated
2. Have an option per client to redirect to some default URL if there is already a session

We are getting a lot of people complaining about this recently.

**Comment by mposolda** [comment]
@stelewis-redhat @alechenninger This is long time issue in Keycloak, which people already complaint a lot in the past. The historic JIRA is here for the reference: https://issues.redhat.com/browse/KEYCLOAK-5179

The core of the problem is, that after authentication in tab1, the "authentication session" is cleared on the server. This means that "tab2" is not able to see authentication session anymore.

We discussed it couple of times how to address this, some of the discussions are in the JIRA, however it is swamped with many comments of users, but there are maybe some interesting ones if you want to take a look. One discussion was back in the old keycloak-dev mailing list: https://lists.jboss.org/pipermail/keycloak-dev/2019-June/012146.html .

We can try to restore the discussion and prioritize this in the next months, but not sure if we have sources for that... You can try to discuss with @stianst about this.

However some workarounds to what you mentioned should be already possible.
1) Note that if your client has the "Base URL" set, then the "You are already logged-in" page will contain the link "Back to client" which user can click and being redirected back to the client.

2) You can try to override the "info.ftl" page of the login theme and possibly do some hack to automatically redirect back to the client. But note that the automatic redirect works fine just for some clients.
For example for the clients like our admin console (security-admin-console client) it works fine because redirect to the client automatically enforce new login, which will happen automatically due the SSO. But for some other clients, which does not enforce automatic login when they are on the "Base URL", there can be tricky situation when instead of the "You are already logged in" page, user will be redirected back to the "Client URL" but not being authenticated. Which is maybe even worse for the users from UXD perspective than showing them "You are already logged-in" page.

**Reply by llatour** [reply]
As stated in this bug: https://github.com/keycloak/keycloak/issues/11746, workarounds are not enough and when using identity provider this leads to other issues.

**Comment by stianst** [comment]
I'd love to see the `You are already logged in` message go away completely and it just doing the right thing to login the user. It's not a trivial thing to achieve though for the reasons @mposolda mentioned, but there most be some way of sorting this out ;)

**Comment by eldarj** [comment]
> The core of the problem is, that after authentication in tab1, the "authentication session" is cleared on the server. This means that "tab2" is not able to see authentication session anymore.

Are we sure this is the case? This doesn't sound right because
1. How come it cannot see the session but it can detect that you are already logged in?
2. Also, all sessions exist and you can see these on the admin console under the session tab, for a given user

I would argue that if we can detect that a user is already logged in, this means we have a session that raises this. This also means that we can assign this tab2 whatever session already exists and thus log in the user automatically on all the other tabs.

**Comment by eldarj** [comment]
I wrote about this on a different **You are already logged in** Github issue, but I'll add this here too.

This page is legit and valid, for whatever reason you end up on the login page, while logged in. You have to look into how your applications integrate with Keycloak and why this occurs.

But my suggestion is, that we improve this in a configurable way, where we provide additionally two options. So we could have something like

- Render **You are already logged** in by default
- Give the option to enable - **Automatic login** through admin console, for a given realm, or client
  - This would mean that Keycloak when encounters an existing user session, just _assigns this_ one to the other tabs. 
  - This covers the _already open tabs_  or whenever a logged in user lands on the login form for whatever reason.
  - Reasoning - your app simply has to log you out properly when you actually do want to login with a different user
- Give the option to enable **Force log out** again through admin console, for a realm/client
  - This would force-logout you if you land back on the login page for whatever reason
  
  The last option might not make sense in some cases, e.g. opening multiple links/tabs from email before authenticating - but for some might still be more suitable rather than rendering the tasteless **You are already logged in**

**Comment by omasseau** [comment]
Indeed an option to be able to at least have a link to redirect to the original request page would be great.
This 'You are already logged in page' is a bit annoying to the users.

**Reply by ghost** [reply]
Hi, @omasseau Do you find any way to do this? I want to redirect the user to original website page.

**Comment by sventorben** [comment]
Hello! What is the current state of this discussion. Is there an open issue related to this or any efforts to change this behavior?

**Reply by ninjaoxygen** [reply]
The main fix is merged here https://github.com/keycloak/keycloak/pull/23517 with some continuing improvements here https://github.com/keycloak/keycloak/pull/23932 and I think https://github.com/keycloak/keycloak/issues/23880, not had chance to test it yet but it sounds good.
